### PR TITLE
platformService tests and move tests from tdd to bdd

### PR DESCRIPTION
**Pull request by ranasaria**
This change contains the following:
1. Tests for platformService
2. Missing Toolsservice tests
2. A test for azdataService access - which weirdly does not work on linux and does work on my mac box. Skipping it for now.
3. Change resource-deployment tests from 'tdd' to 'bdd'
4. Move existing test files to the test folders that mimics src folder structure.

### CODE REVIEW

File: extensions/resource-deployment/src/test/index.ts [hunk 1]
```diff
@@ -9,7 +9,6 @@ const testRunner = require('vscodetestcover');
 const suite = 'resource-deployment Extension Tests';
 
 const mochaOptions: any = {
-	ui: 'tdd',
```
Comment (ranasaria) [inline extensions/resource-deployment/src/test/index.ts:12]: From @Charles-Gagnon on https://github.com/microsoft/azuredatastudio/pull/13183
Let's keep this specific and put 'bdd' for the ones we want to use bdd.

Comment (ranasaria) [reply extensions/resource-deployment/src/test/index.ts:12]: I wanted to use bdd for all of them. I already changed all the existing ones to bdd also. In the future, we can put 'tdd' on the ones that we desire to use 'tdd'

File: extensions/resource-deployment/src/test/services/apiService.test.ts [hunk 2]
```diff
@@ -5,11 +5,15 @@
...
+import { apiService } from '../../services/apiService';
 
-suite('API Service Tests', function (): void {
-	test('getAzurecoreApi returns azure api', () => {
+describe('API Service Tests', function (): void {
+	it('get azurecoreApi returns azure api', () => {
 		const api = apiService.azurecoreApi;
 		assert(api !== undefined);
 	});
+	it.skip('get azdataApi returns azdata api', () => {
```
Comment (ranasaria) [inline extensions/resource-deployment/src/test/services/apiService.test.ts:15]: From @Charles-Gagnon 
Why are you adding a skipped test?

Comment (ranasaria) [reply extensions/resource-deployment/src/test/services/apiService.test.ts:15]: The test works locally on mac but fails for some reason on Linux due to the extension not having been activated when the test runs. Do you have a suggestion on how to force extension activation? Should I remove it?

File: extensions/resource-deployment/src/test/index.ts (hunk 1 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/test/index.ts:12]: changed the value to ui: 'bdd' instead of removing to make it default to 'bdd'

**Review by alanrenmsft — APPROVED** [review]

File: extensions/resource-deployment/src/test/services/apiService.test.ts (hunk 2 above)
Comment (Charles-Gagnon) [reply extensions/resource-deployment/src/test/services/apiService.test.ts:15]: Yeah let's just remove it I think - the test isn't really doing anything useful anyways

File: extensions/resource-deployment/src/test/services/platformService.test.ts [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+import * as sinon from 'sinon';
+import { Readable } from 'stream';
+import * as sudo from 'sudo-prompt';
+import * as TypeMoq from 'typemoq';
+import * as vscode from 'vscode';
+import { tryExecuteAction } from '../../common/utils';
+import { OsDistribution } from '../../interfaces';
+import { extensionOutputChannel, PlatformService } from '../../services/platformService';
+
+class TestChildProcessPromise<T> implements cp.ChildProcessPromise {
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/test/services/platformService.test.ts:21]: This seems like it should probably go in a general stubs.ts (or whatever) module

File: extensions/resource-deployment/src/test/services/platformService.test.ts [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+						(await platformService.fileExists(filePath)).should.be.false();
+						break;
+					}
+					case 'throws': {
+						sinon.stub(fs.promises, 'access').rejects({});
+						const { error } = await tryExecuteAction(() => platformService.fileExists(filePath));
+						should(error).not.be.undefined();
+						break;
+					}
+					default: console.log('unexpected error'); break;
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/test/services/platformService.test.ts:191]: Shouldn't this throw or something?

File: extensions/resource-deployment/src/test/services/platformService.test.ts [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+					sinon.stub(vscode.workspace, 'textDocuments').get(() => [{ isUntitled: true, fileName: '' }]);
+				}
+				const result = platformService.isNotebookNameUsed(title);
+				result.should.equal(isUsed);
+			});
+		}));
+	});
+	describe('runCommand', () => {
+		[
+			//[commandSucceeds, ignoreError]
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/test/services/platformService.test.ts:286]: For this kind of thing I prefer to use objects so you can use named fields : 

```
[
{ commandSucceeds: true, ignoreError: undefined},
...
]
```

File: extensions/resource-deployment/src/test/services/platformService.test.ts [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+});
+
+function verifyCommandExecution(stub: sinon.SinonStub, result: string | undefined, error: any, command: string, commandSucceeds: boolean | undefined, ignoreError: boolean | undefined) {
+	stub.callCount.should.equal(1);
+	if (commandSucceeds) {
+		result!.should.equal(command);
+	} else {
+		if (ignoreError) {
+			should(error).be.undefined();
+			(result === undefined || result.length === 0).should.be.true();
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/test/services/platformService.test.ts:338]: These should have messages saying what is failing - it's not going to display anything useful just by checking true/false

File: extensions/resource-deployment/src/test/services/toolsService.test.ts [hunk 7]
```diff
@@ -42,10 +41,21 @@ suite('Tools Service Tests', function (): void {
...
 		const toolsService = new ToolsService(mockPlatformService.object);
 		const tool = toolsService.getToolByName('no-such-tool');
 		assert(tool === undefined, 'for a not defined tool, expected value is undefined');
 	});
+
+	it('get/set tools for CurrentProvider', () => {
+		const iTools: ITool[] = tools.map(toolInfo => {
+			const tool = toolsService.getToolByName(toolInfo.name);
+			assert(!!tool, `The tool: ${toolInfo.name} is not recognized`);
+			assert.equal(tool!.type, toolInfo.type, 'returned notebook name does not match expected value');
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/test/services/toolsService.test.ts:55]: Use should assertions when you can

**Review by Charles-Gagnon — APPROVED** [review]

File: extensions/resource-deployment/src/test/services/platformService.test.ts (hunk 4 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/test/services/platformService.test.ts:191]: good call. Will change.

File: extensions/resource-deployment/src/test/services/platformService.test.ts (hunk 3 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/test/services/platformService.test.ts:21]: Done.

File: extensions/resource-deployment/src/test/services/platformService.test.ts (hunk 5 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/test/services/platformService.test.ts:286]: Good suggesion. Changed.

File: extensions/resource-deployment/src/test/services/apiService.test.ts (hunk 2 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/test/services/apiService.test.ts:15]: Removed

File: extensions/resource-deployment/src/test/services/platformService.test.ts (hunk 4 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/test/services/platformService.test.ts:191]: Done

File: extensions/resource-deployment/src/test/services/platformService.test.ts (hunk 6 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/test/services/platformService.test.ts:338]: Added message.

File: extensions/resource-deployment/src/test/services/toolsService.test.ts (hunk 7 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/test/services/toolsService.test.ts:55]: Fixed.

**Pull request merged**